canonical/istio-beacon-k8s-operator
# Add metrics integration to istio charms

## Issue
The Istio beacon exposes metrics through a separate workload it deploys:

- **`beacon-waypoint`**: Accessible at `istio-beacon-k8s-{self.model.name}-waypoint.{self.model.name}.svc:15020`, on port `15020` at `/stats/prometheus`.

The main issue with integrating via the beacon's FQDN is that Kubernetes will enforce load balancing when the beacon is scaled, resulting in Prometheus receiving inconsistent metrics.

## Solution
This PR adds metrics integration through the [prometheus_scrape](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) interface, utilizing the newly created [metrics-k8s-proxy](https://github.com/canonical/metrics-k8s-proxy) to aggregate these metrics into Prometheus via a unified endpoint.

## Testing Instructions

1. juju deploy istio-beacon-k8s --trust
2. juju deploy prometheus-k8s --trust
3. juju relate istio-k8s istio-beacon-k8s
4. Explore Prometheus to validate that the metrics are present. Additionally, check for the `up` metric for each the waypoint pod.
